src/h3/frame.rs

```diff
@@ -0,0 +1,879 @@
+        }
+    }
+
+    fn from_bits(bits: u8) -> PrioritizedElemType {
+        match bits {
+            0x00 => PrioritizedElemType::RequestStream,
+            0x01 => PrioritizedElemType::PushStream,
+            0x02 => PrioritizedElemType::Placeholder,
+            0x03 => PrioritizedElemType::CurrentStream,
+            _ => PrioritizedElemType::Error,
+        }
+    }
+}
+
+/// HTTP/3 Element Dependency type.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum ElemDependencyType {
+    RequestStream,
+    PushStream,
+    Placeholder,
+    RootOfTree,
+    Error,
+}
+
+impl ElemDependencyType {
+    fn is_edid_absent(self) -> bool {
+        match self {
+            ElemDependencyType::RootOfTree => true,
+            ElemDependencyType::Error => true,
+            _ => false,
+        }
+    }
+
+    fn to_bits(self) -> u8 {
+        match self {
+            ElemDependencyType::RequestStream => 0x00,
+            ElemDependencyType::PushStream => 0x01,
+            ElemDependencyType::Placeholder => 0x02,
+            ElemDependencyType::RootOfTree => 0x03,
+            _ => 0x04,
+        }
+    }
+
+    fn from_bits(bits: u8) -> ElemDependencyType {
+        match bits {
+            0x00 => ElemDependencyType::RequestStream,
+            0x01 => ElemDependencyType::PushStream,
+            0x02 => ElemDependencyType::Placeholder,
+            0x03 => ElemDependencyType::RootOfTree,
+            _ => ElemDependencyType::Error,
+        }
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub enum Frame {
+    Data {
+        payload: Vec<u8>,
+    },
+
+    Headers {
+        header_block: Vec<u8>,
+    },
+
+    Priority {
+        priority_elem: PrioritizedElemType,
+        elem_dependency: ElemDependencyType,
+        prioritized_element_id: Option<u64>,
+        element_dependency_id: Option<u64>,
+        weight: u8,
+    },
+
+    CancelPush {
+        push_id: u64,
+    },
+
+    Settings {
+        num_placeholders: Option<u64>,
+        max_header_list_size: Option<u64>,
+        qpack_max_table_capacity: Option<u64>,
+        qpack_blocked_streams: Option<u64>,
+    },
+
+    PushPromise {
+        push_id: u64,
+        header_block: Vec<u8>,
+    },
+
+    GoAway {
+        stream_id: u64,
+    },
+
+    MaxPushId {
+        push_id: u64,
+    },
+
+    DuplicatePush {
+        push_id: u64,
+    },
+}
+
+impl Frame {
+    pub fn from_bytes(b: &mut octets::Octets) -> Result<Frame> {
+        let payload_length = b.get_varint()?;
+        let frame_type = b.get_u8()?;
+
+        // debug!("GOT FRAME {:x}, payload_len= {}", frame_type, payload_length);
+
+        // TODO: handling of 0-length frames
+        let frame = match frame_type {
+            DATA_FRAME_TYPE_ID => Frame::Data {
+                payload: b.get_bytes(payload_length as usize)?.to_vec(),
+            },
+
+            HEADERS_FRAME_TYPE_ID => Frame::Headers {
+                header_block: b.get_bytes(payload_length as usize)?.to_vec(),
```

Comment:
Why not just decode QPACK here and put the `Vec<Header>` in the frame? `to_vec()` copies the data, so it's better to avoid it if possible.

Reply:
There are a few weak arguments against doing that. 

First, so far in `quiche::h3:frame`, the structure of frames closely aligns with the HTTP/3 wire format specification. That states that HEADERS frames carry a Header Block i.e. the compressed representation of a header list. Changing that to be a Vec<Header> adds some complexity into this area of code for it to call QPACK. I think things will get worse when we want to use dynamic compression and need to also generate and emit enocoder and decoder instructions.

The other argument against embedding the compression inside, is that it gives us greater flexibility to change header compression. I'm not convinced we need that flexibility right now.



1) It couples the HTTP/3 wire format serilization and deserialization to one QPACK li